### PR TITLE
Feature/nullable query params

### PR DESCRIPTION
**Pull request by serweck**
Fix nullable parameters in Query Params

Fix Issues:
https://github.com/Xabaril/Acheve.TestHost/issues/47
https://github.com/Xabaril/Acheve.TestHost/issues/29

### CODE REVIEW

File: src/Acheve.TestHost/Acheve.TestHost.csproj [hunk 1]
```diff
@@ -15,7 +15,7 @@
   
   <ItemGroup>
     <Folder Include="Routing\AttributeTemplates\" />
-    <Folder Include="Routing\Tokenizers\" />
+	<Folder Include="Routing\Tokenizers\" />
```
Comment (Sergio1192) [inline src/Acheve.TestHost/Acheve.TestHost.csproj:18]: wrong identation

Comment (serweck) [reply src/Acheve.TestHost/Acheve.TestHost.csproj]: fixed

**Comment by serweck** [comment]
I think now is find after merge develop

File: src/Acheve.TestHost/Routing/Tokenizers/PrimitiveTypeHelper.cs [hunk 2]
```diff
@@ -0,0 +1,17 @@
+﻿using System;
+
+namespace Acheve.TestHost.Routing.Tokenizers
+{
+    public static class PrimitiveTypeHelper
```
Comment (Sergio1192) [inline src/Acheve.TestHost/Routing/Tokenizers/PrimitiveTypeHelper.cs:5]: Change the name to TypeExtensions

Comment (Sergio1192) [inline src/Acheve.TestHost/Routing/Tokenizers/PrimitiveTypeHelper.cs:5]: Maybe internal?

Comment (serweck) [reply src/Acheve.TestHost/Routing/Tokenizers/PrimitiveTypeHelper.cs]: done

File: src/Acheve.TestHost/Routing/Tokenizers/PrimitiveTypeHelper.cs [hunk 3]
```diff
@@ -0,0 +1,17 @@
+﻿using System;
+
+namespace Acheve.TestHost.Routing.Tokenizers
```
Comment (Sergio1192) [inline src/Acheve.TestHost/Routing/Tokenizers/PrimitiveTypeHelper.cs]: Use "System" namespace

**Comment by Sergio1192** [comment]
Please, add the comments in HttpResponseMessageExtensions/ReadContentAsAsync. I didn't notice before.

File: src/Acheve.TestHost/Routing/Tokenizers/TypeExtensions.cs [hunk 4]
```diff
@@ -0,0 +1,17 @@
+﻿using System;
+
+namespace Acheve.TestHost.Routing.Tokenizers
```
Comment (Sergio1192) [inline src/Acheve.TestHost/Routing/Tokenizers/TypeExtensions.cs:3]: Use "System" namespace

File: src/Acheve.TestHost/Routing/Tokenizers/PrimitiveParameterActionTokenizer.cs [hunk 5]
```diff
@@ -16,31 +16,29 @@ public void AddTokens<TController>(TestServerAction action, TestServerTokenColle
...
-                    }
-                    else if (parameters[i].ParameterType.IsArray
-                       && IsPrimitiveType(parameters[i].ParameterType.GetElementType()))
+                    if (tokenValue != null)
                     {
-                        var arrayValues = (Array)action.ArgumentValues[i].Instance;
+                        tokens.AddToken(tokenName, tokenValue.ToString(), isConventional: false);
+                    }
+                }
+                else if (parameters[i].ParameterType.IsArray
```
Comment (Sergio1192) [inline src/Acheve.TestHost/Routing/Tokenizers/PrimitiveParameterActionTokenizer.cs:30]: why do u remove "!IgnoreHeader" in this case?

File: src/Acheve.TestHost/Routing/Tokenizers/PrimitiveParameterActionTokenizer.cs [hunk 6]
```diff
@@ -16,31 +16,29 @@ public void AddTokens<TController>(TestServerAction action, TestServerTokenColle
...
             {
-                if (!IgnoreHeader(parameters[i]))
+                if (parameters[i].ParameterType.IsPrimitiveType()
+                    && !IgnoreHeader(parameters[i]))
                 {
-                    if (IsPrimitiveType(parameters[i].ParameterType))
-                    {
-                        var tokenName = parameters[i].Name.ToLowerInvariant();
-                        var tokenValue = action.ArgumentValues[i].Instance;
+                    var tokenName = parameters[i].Name.ToLowerInvariant();
```
Comment (Sergio1192) [inline src/Acheve.TestHost/Routing/Tokenizers/PrimitiveParameterActionTokenizer.cs:22]: Maybe you could extract this variable outside and share with all cases

File: src/Acheve.TestHost/Routing/Tokenizers/ComplexParameterActionTokenizer.cs [hunk 7]
```diff
@@ -15,9 +16,10 @@ class ComplexParameterActionTokenizer
             for (int i = 0; i < parameters.Length; i++)
             {
                 var type = parameters[i].ParameterType;
-                var instance = action.ArgumentValues[i].Instance;
+                var instance = action.ArgumentValues.Any(x => x.Key == i) ? action.ArgumentValues[i].Instance : null;
 
-                if (instance != null && !(type.IsPrimitive || type == typeof(String) || type == typeof(Decimal) || type == typeof(Guid)))
+
```
Comment (Sergio1192) [inline src/Acheve.TestHost/Routing/Tokenizers/ComplexParameterActionTokenizer.cs:21]: Remove this extra space

File: src/Acheve.TestHost/Routing/Tokenizers/PrimitiveParameterActionTokenizer.cs (hunk 5 above)
Comment (serweck) [reply src/Acheve.TestHost/Routing/Tokenizers/PrimitiveParameterActionTokenizer.cs:30]: was a mistake. Resolved

File: src/Acheve.TestHost/Routing/Tokenizers/TypeExtensions.cs (hunk 4 above)
Comment (serweck) [reply src/Acheve.TestHost/Routing/Tokenizers/TypeExtensions.cs:3]: done

File: src/Acheve.TestHost/Routing/Tokenizers/PrimitiveParameterActionTokenizer.cs (hunk 6 above)
Comment (serweck) [reply src/Acheve.TestHost/Routing/Tokenizers/PrimitiveParameterActionTokenizer.cs:22]: done

File: src/Acheve.TestHost/Routing/Tokenizers/ComplexParameterActionTokenizer.cs (hunk 7 above)
Comment (serweck) [reply src/Acheve.TestHost/Routing/Tokenizers/ComplexParameterActionTokenizer.cs:21]: done

File: src/Acheve.TestHost/HttpResponseMessageExtensions.cs [hunk 8]
```diff
@@ -22,6 +22,12 @@ public static async Task IsSuccessStatusCodeOrThrow(this HttpResponseMessage res
             throw new Exception($"Response status does not indicate success: {response.StatusCode:D} ({response.StatusCode}); \r\n{content}");
         }
 
+        /// <summary>
+        /// Read HttpResponseMessage and convert to T Class
+        /// </summary>
+        /// <typeparam name="T">Class</typeparam>
+        /// <param name="responseMessage">The httpResponseMessage instance</param>
+        /// <returns></returns>
```
Comment (Sergio1192) [inline src/Acheve.TestHost/HttpResponseMessageExtensions.cs:30]: It returns something

File: src/Acheve.TestHost/Routing/Tokenizers/PrimitiveParameterActionTokenizer.cs (hunk 5 above)
Comment (Sergio1192) [reply src/Acheve.TestHost/Routing/Tokenizers/PrimitiveParameterActionTokenizer.cs:32]: In preview version, you set one if with ignoreHeader and after the other ifs, I prefer this preview version

Comment (serweck) [reply src/Acheve.TestHost/Routing/Tokenizers/PrimitiveParameterActionTokenizer.cs]: this version ?

File: src/Acheve.TestHost/HttpResponseMessageExtensions.cs (hunk 8 above)
Comment (serweck) [reply src/Acheve.TestHost/HttpResponseMessageExtensions.cs]: done

**Review by Sergio1192 — APPROVED** [review]

**Pull request merged**